### DISCUSSION TITLE
Can the local mouse pointer cursor be synchronized with the controlled host?

### DISCUSSION THREAD

**Discussion by yaole-A** [ClassicOldSong/Apollo#80] [General]
In a non-LAN environment, the mouse delay is particularly severe, which is particularly uncomfortable when playing League of Legends and also very uncomfortable when on the computer desktop. Of course, the local mouse mode (press CTRL ALT SHIFT C) can solve it, but the mouse cursor will not change. It is okay on the desktop, but it is a bit uncomfortable in League of Legends. Can it get the information of the remote host cursor like parsec, and then draw locally? Synchronize the remote cursor to the local

**Comment by ClassicOldSong** [comment]
This is a planned feature. Will implement in the future.

**Reply by boilfish** [reply]
It looks like CTRL+ALT+SHIFT+N can not hide windows white cursor in lastest version. anything changed?

Artemis android 241031(bluetooth mouse) -> Apollo windows10 v0.2.6-alpha.2 Virtual Display. There would be a white windows host cursor chasing the black android native cursor. When press CTRL+ALT+SHIFT+N the white one didn't hide. In my mind it has worked before.

**Reply by ClassicOldSong** [reply]
Nothing changed and it's still working on my end(Windows 11). Windows might did something strange in the mean time.

**Reply by boilfish** [reply]
oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor?
if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~~Or maybe I'm remembering it wrong？~~

edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong

**Reply by ClassicOldSong** [reply]
Apollo handles this correctly on my end and I've tested through multiple clients/hosts. If it's not working, it might be Windows 10 can't handle hardware cursor for virtual displays.

**Reply by ClassicOldSong** [reply]
I think I have said it clearly: Windows 10 is very likely not able to handle hardware cursor correctly on virtual displays. With Sunshine, you're streaming from physical display, and Windows thinks there's hardware cursor support so it won't draw the cursor on the screen before sending the image for capturing/display, but with virtual displays, Windows don't think there's hardware cursor support on them so it's always drawing the cursor on the screen for you.

On Windows 11 this problem never happened.

**Reply by boilfish** [reply]
> I think I have said it clearly: Windows 10 is very likely not able to handle hardware cursor correctly on virtual displays. With Sunshine, you're streaming from physical display, and Windows thinks there's hardware cursor support so it won't draw the cursor on the screen before sending the image for capturing/display, but with virtual displays, Windows don't think there's hardware cursor support on them so it's always drawing the cursor on the screen for you.
> 
> On Windows 11 this problem never happened.

Got it. Thank you! maybe time to update to win11😂

**Reply by yaole-A** [reply]
> oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor? if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~Or maybe I'm remembering it wrong？~-----------谷歌翻译-----------哦，我在 Windows 上尝试了 CTRL+ALT+SHIFT+N，Android 黑色光标消失了。有没有办法隐藏windows光标？如果从 Artemis 一侧发送 CTRL+ALT+SHIFT+N 到阳光，白色 Windows 光标将隐藏。但阿波罗没有。还是我记错了？
> 
> edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong-----------谷歌翻译-----------编辑：我重新安装了sunshine v0.23.1并进行了测试。确实如此。 @经典老歌

In which case did you press ctrl alt shift N? Only when the client sends this command can the mouse pointer of the controlled end be hidden. Pressing this command on the host side has no effect. Sunshine does not listen to shortcut keys locally.

**Reply by boilfish** [reply]
> > oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor? if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~Or maybe I'm remembering it wrong？~-----------谷歌翻译-----------哦，我在 Windows 上尝试了 CTRL+ALT+SHIFT+N，Android 黑色光标消失了。有没有办法隐藏windows光标？如果从 Artemis 一侧发送 CTRL+ALT+SHIFT+N 到阳光，白色 Windows 光标将隐藏。但阿波罗没有。还是我记错了？
> > edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong-----------谷歌翻译-----------编辑：我重新安装了sunshine v0.23.1并进行了测试。确实如此。 @经典老歌
> 
> In which case did you press ctrl alt shift N? Only when the client sends this command can the mouse pointer of the controlled end be hidden. Pressing this command on the host side has no effect. Sunshine does not listen to shortcut keys locally.

i don't know why, but it work🫠. the black android cursor disappear until i move the android's Bluetooth mouse.

**Reply by yaole-A** [reply]
> > > oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor? if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~Or maybe I'm remembering it wrong？~-----------谷歌翻译-----------哦，我在 Windows 上尝试了 CTRL+ALT+SHIFT+N，Android 黑色光标消失了。有没有办法隐藏windows光标？如果从 Artemis 一侧发送 CTRL+ALT+SHIFT+N 到阳光，白色 Windows 光标将隐藏。但阿波罗没有。还是我记错了？
> > > edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong-----------谷歌翻译-----------编辑：我重新安装了sunshine v0.23.1并进行了测试。确实如此。 @经典老歌
> > 
> > 
> > In which case did you press ctrl alt shift N? Only when the client sends this command can the mouse pointer of the controlled end be hidden. Pressing this command on the host side has no effect. Sunshine does not listen to shortcut keys locally.
> 
> i don't know why, but it work🫠. the black android cursor disappear until i move the android's Bluetooth mouse.-----------谷歌翻译-----------我不知道为什么，但它有效。黑色的 Android 光标消失，直到我移动 Android 的蓝牙鼠标。

So it is not a system problem of win10 and win11, nor is it a problem of apollo. only the shortcut keys you send in moonlight will  take effect.No matter how you think about it, the shortcut keys you press in host cannot control the hiding or display of the Android mouse cursor.Have you ever thought that the Android cursor disappears because you don't move for a bit time and it disappears by itself?

**Reply by boilfish** [reply]
> > > > oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor? if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~Or maybe I'm remembering it wrong？~-----------谷歌翻译-----------哦，我在 Windows 上尝试了 CTRL+ALT+SHIFT+N，Android 黑色光标消失了。有没有办法隐藏windows光标？如果从 Artemis 一侧发送 CTRL+ALT+SHIFT+N 到阳光，白色 Windows 光标将隐藏。但阿波罗没有。还是我记错了？
> > > > edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong-----------谷歌翻译-----------编辑：我重新安装了sunshine v0.23.1并进行了测试。确实如此。 @经典老歌
> > > 
> > > 
> > > In which case did you press ctrl alt shift N? Only when the client sends this command can the mouse pointer of the controlled end be hidden. Pressing this command on the host side has no effect. Sunshine does not listen to shortcut keys locally.
> > 
> > 
> > i don't know why, but it work🫠. the black android cursor disappear until i move the android's Bluetooth mouse.-----------谷歌翻译-----------我不知道为什么，但它有效。黑色的 Android 光标消失，直到我移动 Android 的蓝牙鼠标。
> 
> So it is not a system problem of win10 and win11, nor is it a problem of apollo. only the shortcut keys you send in moonlight will take effect.No matter how you think about it, the shortcut keys you press in host cannot control the hiding or display of the Android mouse cursor.Have you ever thought that the Android cursor disappears because you don't move for a bit time and it disappears by itself?

no, the android cursor disappear immediately when i press the combo in host. I still think it is a bug of win10 virtual display because the maintainer has tested the combo work well on win11 virtual display mode. And i'v tested this combo is work on win10 if i choose legacy mode without virtual display.

**Reply by yaole-A** [reply]
> > > oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor? if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~Or maybe I'm remembering it wrong？~-----------谷歌翻译-----------哦，我在 Windows 上尝试了 CTRL+ALT+SHIFT+N，Android 黑色光标消失了。有没有办法隐藏windows光标？如果从 Artemis 一侧发送 CTRL+ALT+SHIFT+N 到阳光，白色 Windows 光标将隐藏。但阿波罗没有。还是我记错了？
> > > edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong-----------谷歌翻译-----------编辑：我重新安装了sunshine v0.23.1并进行了测试。确实如此。 @经典老歌
> > 
> > 
> > In which case did you press ctrl alt shift N? Only when the client sends this command can the mouse pointer of the controlled end be hidden. Pressing this command on the host side has no effect. Sunshine does not listen to shortcut keys locally.
> 
> i don't know why, but it work🫠. the black android cursor disappear until i move the android's Bluetooth mouse.-----------谷歌翻译-----------我不知道为什么，但它有效。黑色的 Android 光标消失，直到我移动 Android 的蓝牙鼠标。

I just understand you.I remembered that if it is a virtual display, the win10 system cannot handle the hardware cursor of the virtual display device. To realize the function of hiding the cursor, your display must support the hardware mouse. All the physical display actually supports hardware cursor. I don’t know whether the virtual display currently used by Apollo supports the hardware pointer. But no matter which open source virtual display driver you use, make sure your system is win11. Only win11 supports the hardware mouse of the virtual reality device.if you don't want to use win11, using physical dsiplay is the solution.

**Reply by yaole-A** [reply]
> > > > > oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor? if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~Or maybe I'm remembering it wrong？~-----------谷歌翻译-----------哦，我在 Windows 上尝试了 CTRL+ALT+SHIFT+N，Android 黑色光标消失了。有没有办法隐藏windows光标？如果从 Artemis 一侧发送 CTRL+ALT+SHIFT+N 到阳光，白色 Windows 光标将隐藏。但阿波罗没有。还是我记错了？
> > > > > edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong-----------谷歌翻译-----------编辑：我重新安装了sunshine v0.23.1并进行了测试。确实如此。 @经典老歌
> > > > 
> > > > 
> > > > In which case did you press ctrl alt shift N? Only when the client sends this command can the mouse pointer of the controlled end be hidden. Pressing this command on the host side has no effect. Sunshine does not listen to shortcut keys locally.
> > > 
> > > 
> > > i don't know why, but it work🫠. the black android cursor disappear until i move the android's Bluetooth mouse.-----------谷歌翻译-----------我不知道为什么，但它有效。黑色的 Android 光标消失，直到我移动 Android 的蓝牙鼠标。
> > 
> > 
> > So it is not a system problem of win10 and win11, nor is it a problem of apollo. only the shortcut keys you send in moonlight will take effect.No matter how you think about it, the shortcut keys you press in host cannot control the hiding or display of the Android mouse cursor.Have you ever thought that the Android cursor disappears because you don't move for a bit time and it disappears by itself?
> 
> no, the android cursor disappear immediately when i press the combo in host. I still think it is a bug of win10 virtual display because the maintainer has tested the combo work well on win11 virtual display mode. And i'v tested this combo is work on win10 if i choose legacy mode without virtual display.-----------谷歌翻译-----------不，当我按下主机中的组合键时，Android 光标立即消失。我仍然认为这是win10虚拟显示的一个错误，因为维护者已经测试了该组合在win11虚拟显示模式下工作良好。我测试过如果我选择没有虚拟显示的传统模式，这个组合可以在 win10 上工作。

It's not a bug, it's that the win10 system does not support the hardware mouse of the virtual display device.If you want to use win10, you can't hide the mouse with the virtual display. only  the physical display can you hide it in win10.by the way,the sunshine host does not listen to the local mouse and keyboard operations, how can it control your other end

**Reply by boilfish** [reply]
> > > > > > oh, i tried CTRL+ALT+SHIFT+N on windows and the android black cursor disappear. is there anyway to hide windows cursor? if send CTRL+ALT+SHIFT+N from Artemis side to sunshine the white Windows cursor will hide. but Apollo didn't. ~Or maybe I'm remembering it wrong？~-----------谷歌翻译-----------哦，我在 Windows 上尝试了 CTRL+ALT+SHIFT+N，Android 黑色光标消失了。有没有办法隐藏windows光标？如果从 Artemis 一侧发送 CTRL+ALT+SHIFT+N 到阳光，白色 Windows 光标将隐藏。但阿波罗没有。还是我记错了？
> > > > > > edit: I reinstalled sunshine v0.23.1 and tested. It really is. @ClassicOldSong-----------谷歌翻译-----------编辑：我重新安装了sunshine v0.23.1并进行了测试。确实如此。 @经典老歌
> > > > > 
> > > > > 
> > > > > In which case did you press ctrl alt shift N? Only when the client sends this command can the mouse pointer of the controlled end be hidden. Pressing this command on the host side has no effect. Sunshine does not listen to shortcut keys locally.
> > > > 
> > > > 
> > > > i don't know why, but it work🫠. the black android cursor disappear until i move the android's Bluetooth mouse.-----------谷歌翻译-----------我不知道为什么，但它有效。黑色的 Android 光标消失，直到我移动 Android 的蓝牙鼠标。
> > > 
> > > 
> > > So it is not a system problem of win10 and win11, nor is it a problem of apollo. only the shortcut keys you send in moonlight will take effect.No matter how you think about it, the shortcut keys you press in host cannot control the hiding or display of the Android mouse cursor.Have you ever thought that the Android cursor disappears because you don't move for a bit time and it disappears by itself?
> > 
> > 
> > no, the android cursor disappear immediately when i press the combo in host. I still think it is a bug of win10 virtual display because the maintainer has tested the combo work well on win11 virtual display mode. And i'v tested this combo is work on win10 if i choose legacy mode without virtual display.-----------谷歌翻译-----------不，当我按下主机中的组合键时，Android 光标立即消失。我仍然认为这是win10虚拟显示的一个错误，因为维护者已经测试了该组合在win11虚拟显示模式下工作良好。我测试过如果我选择没有虚拟显示的传统模式，这个组合可以在 win10 上工作。
> 
> It's not a bug, it's that the win10 system does not support the hardware mouse of the virtual display device.If you want to use win10, you can't hide the mouse with the virtual display. only the physical display can you hide it in win10

Thank you! I understand.

**Comment by yaole-A** [comment]
> This is a planned feature. Will implement in the future.

The project （link：https://github.com/cryonox/moonlight_mousesync ） can synchronize the desktop cursor with the remote, but the cursor in the game cannot. This method is relatively simple and not suitable for games. For the desktop, it is already very good. Parsec's mouse pointer has always been very good, which is why I don't abandon it. Sunshine This problem has been around for a long time, but the developers don't seem to pay attention to it.  looking forward to your addition of this function in the future.

**Comment by yaole-A** [comment]
Maybe the DXGI API can be used to retrieve the game mouse cursor information from the frame data? Most games use hardware mouse pointers (overlaid on top of the screen frame buffer), and I think this might be captured from the rendering layer.For games like Baldur's Gate 3 and League of Legends (hoping the local mouse mode cursor can sync with the cursor in the game as well).

**Reply by ClassicOldSong** [reply]
Sunshine already did this and it composites the mouse onto the frame again before encoding the frame, but the code is a total mess so it's really hard to get the cursor image out separately.

**Reply by yaole-A** [reply]
> Sunshine already did this and it composites the mouse onto the frame again before encoding the frame, but the code is a total mess so it's really hard to get the cursor image out separately.-----------谷歌翻译-----------Sunshine 已经这样做了，它在对帧进行编码之前再次将鼠标合成到帧上，但代码完全混乱，因此很难单独取出光标图像。

em，wish you can achieve it，How long is this plan expected to take to realize?

**Comment by kodidu** [comment]
Hello!

Any update on this cursor icon sync topic?
